File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let dieScore die =
    match die with
    | One -> 1
    | Two -> 2
    | Three -> 3
    | Four -> 4
    | Five -> 5
    | Six -> 6

let sum dice target =
    dice
    |> List.map dieScore
    |> List.filter (fun score -> score = target)
    |> List.sum
    
let fullHouse dice =
    let groups = dice |> List.groupBy id
    
    let hasThreeOfAKind = groups |> List.exists (fun (_, group) -> group.Length = 3)
    let hasTwoOfAKind = groups |> List.exists (fun (_, group) -> group.Length = 2)

    if (hasThreeOfAKind && hasTwoOfAKind)
    then dice |> List.sumBy dieScore
    else 0

let fourOfAKind dice =
    dice
    |> List.groupBy id
    |> List.filter (fun (_, group) -> group.Length >= 4)
    |> List.map snd
    |> List.collect id
    |> List.truncate 4
    |> List.sumBy dieScore

let littleStraight dice =
    let orderedDice =
        dice
        |> List.sortBy dieScore

    match orderedDice with
    | [One; Two; Three; Four; Five] -> 30
    | _ -> 0

let bigStraight dice =
    let orderedDice =
        dice
        |> List.sortBy dieScore

    match orderedDice with
    | [Two; Three; Four; Five; Six] -> 30
    | _ -> 0

let choice dice =
    dice
    |> List.map dieScore
    |> List.sum

let yacht dice =
    let isFiveOfAKind =
        dice
        |> List.groupBy id
        |> List.exists (fun (_, group) -> group.Length = 5)

    if isFiveOfAKind then 50 else 0
    
let score category dice =
    match category with
    | Ones -> sum dice 1
    | Twos -> sum dice 2
    | Threes -> sum dice 3
    | Fours -> sum dice 4
    | Fives -> sum dice 5
    | Sixes -> sum dice 6
    | FullHouse -> fullHouse dice
    | FourOfAKind -> fourOfAKind dice
    | LittleStraight -> littleStraight dice
    | BigStraight -> bigStraight dice
    | Choice -> choice dice
    | Yacht -> yacht dice

File: csharp/yacht/Iteration1/Yacht.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{
    public static int Score(int[] dice, YachtCategory category) =>
        category switch
        {
             YachtCategory.Ones           => dice.Tally(1),
             YachtCategory.Twos           => dice.Tally(2),
             YachtCategory.Threes         => dice.Tally(3),
             YachtCategory.Fours          => dice.Tally(4),
             YachtCategory.Fives          => dice.Tally(5),
             YachtCategory.Sixes          => dice.Tally(6),
             YachtCategory.FullHouse      => dice.FullHouse(),
             YachtCategory.FourOfAKind    => dice.FourOfAKind(),
             YachtCategory.LittleStraight => dice.LittleStraight(),
             YachtCategory.BigStraight    => dice.BigStraight(),
             YachtCategory.Choice         => dice.Choice(),
             YachtCategory.Yacht          => dice.Yacht(),
             _                            => 0
        };
}

public static class YachtGameExtensions
{
    public static int Tally(this IEnumerable<int> dice, int number) =>
        dice
            .Where(die => die == number)
            .Sum();

    public static int FullHouse(this IEnumerable<int> dice)
    {
        var groups = dice.GroupBy(die => die);
        var hasTwoOfAKind = groups.Any(g => g.Count() == 2);
        var hasThreeOfAKind = groups.Any(g => g.Count() == 3);

        if (hasTwoOfAKind && hasThreeOfAKind) return dice.Sum();
        return 0;
    }

    public static int Yacht(this IEnumerable<int> dice) =>
        dice
            .GroupBy(die => die)
            .Where(g => g.Count() == 5)
            .Select(g => 50)
            .DefaultIfEmpty(0)
            .First();

    public static int Choice(this IEnumerable<int> dice) =>
        dice.Sum();

    public static int FourOfAKind(this IEnumerable<int> dice) =>
        dice
            .GroupBy(die => die)
            .Where(g => g.Count() >= 4)
            .SelectMany(g => g)
            .Take(4)
            .Sum();

    public static int LittleStraight(this IEnumerable<int> dice) {
        if (dice.OrderBy(die => die).ToList() is [1, 2, 3, 4, 5]) return 30;
        return 0;
    }

    public static int BigStraight(this IEnumerable<int> dice) {
        if (dice.OrderBy(die => die).ToList() is [2, 3, 4, 5, 6]) return 30;
        return 0;
    }
}